storage: Expose replicated-only distinction in replicaDataIterator

nvb opened the pull request:
Fixes #4572.

This allowed us to simplify some logic in `Replica.Snapshot` and
`Replica.applySnapshot`. It also allowed us to avoid scanning 
unreplicated keys in the GC Queue, at the expense of not being 
able to set the last verification timestamp at the end of GC processing.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/4960)

<!-- Reviewable:end -->

nvb committed:
storage: Expose replicated-only distinction in replicaDataIterator

Fixes #4572.

This allowed us to simplify some logic in `Replica.Snapshot` and
`Replica.applySnapshot`. It also introduced the question of scanning
over unreplicated in the GC queue.

bdarnell commented on storage/replica_raftstorage.go:
```diff
@@ -363,7 +363,7 @@ func (r *Replica) Snapshot() (raftpb.Snapshot, error) {
 	// Iterate over all the data in the range, including local-only data like
 	// the sequence cache.
-	iter := newReplicaDataIterator(&desc, snap)
+	iter := newReplicaDataIterator(&desc, snap, false /* !replicatedOnly */)
```
Generating snapshots should also be limited to replicated data.

bdarnell commented on storage/replica_raftstorage.go:
```diff
@@ -520,19 +507,6 @@ func (r *Replica) applySnapshot(batch engine.Engine, snap raftpb.Snapshot) (uint
 		}
```
While iterating over `snapData.KV` we should also drop any unreplicated keys. This is an extra layer of safety against a snapshot that includes unreplicated data, so it won't actually do anything once you make the change in my previous comment. But I think this change as it stands is incorrect since the HardState will be overwritten by the one from the snapshot. This suggests a gap in our test coverage, but I'm not sure how exactly you would engineer a situation in which transferring HardState across snapshots would actually make a difference.

nvanbenschoten commented:
Yeah I didn't think this was 100% correct, but it was the only way I could get tests passing so I wanted to get some feedback. The issue I'm running into is that when I only add replicated keys to the snapshot, none of the raft log is included. This causes `TestStoreRangeMergeNonCollocated` to fail with the following error.

```
panic: requested entry at index is unavailable

goroutine 2592 [running]:
panic(0x50e8ec0, 0xc820206200)
	/usr/local/go/src/runtime/panic.go:464 +0x3e6
github.com/coreos/etcd/raft.(*raftLog).term(0xc8201858f0, 0x15, 0x0, 0x0, 0x0)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/log.go:241 +0x1a9
github.com/coreos/etcd/raft.(*raftLog).matchTerm(0xc8201858f0, 0x15, 0x6, 0xc820294a80)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/log.go:275 +0x2b
github.com/coreos/etcd/raft.(*raftLog).maybeAppend(0xc8201858f0, 0x15, 0x6, 0x15, 0xc82023e7d0, 0x1, 0x1, 0x16, 0x0)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/log.go:78 +0x5d
github.com/coreos/etcd/raft.(*raft).handleAppendEntries(0xc8201ce540, 0x3, 0x2, 0x1, 0x6, 0x6, 0x15, 0xc82023e7d0, 0x1, 0x1, ...)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/raft.go:761 +0x1fa
github.com/coreos/etcd/raft.stepFollower(0xc8201ce540, 0x3, 0x2, 0x1, 0x6, 0x6, 0x15, 0xc82023e7d0, 0x1, 0x1, ...)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/raft.go:732 +0x3ad
github.com/coreos/etcd/raft.(*raft).Step(0xc8201ce540, 0x3, 0x2, 0x1, 0x6, 0x6, 0x15, 0xc82023e7d0, 0x1, 0x1, ...)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/raft.go:564 +0x3e0
github.com/coreos/etcd/raft.(*RawNode).Step(0xc8202a6700, 0x3, 0x2, 0x1, 0x6, 0x6, 0x15, 0xc82023e7d0, 0x1, 0x1, ...)
	/Users/nathan/Go/src/github.com/coreos/etcd/raft/rawnode.go:175 +0x19d
github.com/cockroachdb/cockroach/storage.(*Store).handleRaftMessage(0xc820188600, 0xc8201d3b00, 0x0, 0x0)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/storage/store.go:1754 +0x275
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/storage/store.go:1811 +0x525
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8201843f0, 0xc820178e50)
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:98 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
	/Users/nathan/Go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:99 +0x62
```

It seems like the tests pass when any of the other "unreplicated" keys are omitted from the snapshot, but when the raft log keys are removed, the issue appears. I'm not deleting the raft log when applying the snapshot, so I'm not exactly sure what's going on. 

Is this expected behavior? I know you had mentioned that most of the raft log *is* actually replicated, but the most recent keys may not be consistent. Should we be keeping track of the replicated vs. unreplicated raft log keys?

bdarnell commented on storage/gc_queue.go:
```diff
@@ -150,7 +150,7 @@ func (gcq *gcQueue) process(now roachpb.Timestamp, repl *Replica,
 	snap := repl.store.Engine().NewSnapshot()
 	desc := repl.Desc()
-	iter := newReplicaDataIterator(desc, snap)
+	iter := newReplicaDataIterator(desc, snap, false /* !replicatedOnly */)
```
The GC queue will never do anything with any unreplicated data, so we could pass true here and make the GC scan a bit less expensive (mainly by ignoring the raft log). However, if we did so the GC queue would no longer be a superset of the verification queue and it would no longer be appropriate to set the verification timestamp at the end of the GC queue's run (see #4710). 

I don't have a good enough sense of how often the GC queue runs to tell whether it's a net win to make GC scans cheaper at the expense of more frequent verification scans. However, if we leave this set to false it should at least be commented that we are doing so to ensure that this doubles as a verification scan. (Maybe `replicaDataIterator` should be responsible for setting the verification timestamp instead of the queues, so it can be set whenever a full iteration has completed but not for partial ones)

bdarnell commented:
Yeah, the raft log is a grey area for replicated/unreplicated. In our current design the log needs to be transferred with the snapshot (at least the portion of the log between firstIndex and appliedIndex). I think it would be best to construct the snapshot with a `replicaDataIterator` that uses only replicated data, and then load the log entries separately with `r.Entries(r.firstIndex, r.appliedIndex)`. You could either reconstruct key-value pairs to insert them into the existing `RaftSnapshotData` proto or just add a separate `repeated raftpb.Entry` field to the proto.

An alternate design that would remove the need to transfer the log with the snapshot would be to make `firstIndex` and (and `RaftTruncatedStateKey` which stores it) unreplicated, so that each replica can set its own `firstIndex` based on the snapshot it received. However, pulling on this thread would require larger changes to the way we do log truncation (each replica would need to be able to independently truncate its log instead of relying on a raft command from the leader)

bdarnell commented on storage/replica_raftstorage.go:
```diff
@@ -484,24 +484,11 @@ func (r *Replica) applySnapshot(batch engine.Engine, snap raftpb.Snapshot) (uint
 
 	// Delete everything in the range and recreate it from the snapshot.
-	iter := newReplicaDataIterator(&desc, batch)
+	iter := newReplicaDataIterator(&desc, batch, true /* replicatedOnly */)
```
We need to delete any old log entries here: any log entries that predate the snapshot will be orphaned and never truncated or GC'd. It's probably better to continue to treat the HardState as a special case; it should be the only thing that needs to be preserved across `applySnapshot`; the other unreplicated state can be discarded.

nvanbenschoten commented:
Based off of what I'm seeing right now, it looks like more than just the raft log between firstIndex and appliedIndex needs to be transferred for the "requested entry at index" to be is unavailable. The only way I have been able to get the tests to pass is to load the Raft Log entries using `r.Entries(r.firstIndex, r.appliedIndex+1)` or to pass unreplicated data in the snapshot KV pairs as we were doing before. I'm not sure why the `appliedIndex` needs to be incremented, but I'm wondering if that means we're counting ourselves. Do you have any idea what could be causing this, and if incrementing the index is correct or a hack that shouldnt be working?

bdarnell commented:
That makes sense. `Entries()` returns a half-open interval `[lo, hi)`, so to include `appliedIndex` in the records to be transferred you have to call `Entries(firstIndex, appliedIndex+1)`.

LGTM

---

Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[roachpb/internal_raft.proto, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSFdrVkFc7IaOlDU29:-KCSFdrVkFc7IaOlDU2A:-1525216466) ([raw file](https://github.com/cockroachdb/cockroach/blob/a88d582d39703966f9a8dfbdf3c812d8dd49ceda/roachpb/internal_raft.proto#L63)):*
Why did these all move? Are there any differences besides this new field?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-:-KCSF9cMegDxcf9PkWIx:-944536391)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[roachpb/internal_raft.proto, line 63 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSFdrVkFc7IaOlDU29:-KCSKV9mVtSS1eyn5uV9:-1596165871) ([raw file](https://github.com/cockroachdb/cockroach/blob/a88d582d39703966f9a8dfbdf3c812d8dd49ceda/roachpb/internal_raft.proto#L63)):*
The only difference is the new field. The reason for this change is that we have now established a dependency on `raftpb`. As @tamird pointed out, we shouldn't need any dependencies on `raftpb` for any protos that need C++ generated code (which is why we dont even link to CoreOS when compiling the C++ protos). This implied that there was a natural separation between what's in this new file and what's in the old file.

---

*[storage/gc_queue.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCISF9lv91JoOmmvNHD-r1-153:-KCSMpD1cgjNO5ZWwDMy:-1471223627) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf51391dcda579fe83757e32ef9443ac0b32a176/storage/gc_queue.go#L153)):*
That's an interesting point, and something I'd be curious to look into in the future. I've added a comment for now.

@spencerkimball, do you have any insight on the tradeoffs here?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[storage/gc_queue.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCISF9lv91JoOmmvNHD-r1-153:-KCSOwH-Rh9A4UuDGzeA:1860085460) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf51391dcda579fe83757e32ef9443ac0b32a176/storage/gc_queue.go#L153)):*
You should absolutely switch to using `replicatedOnly` on this iter and remove the update call to setLastVerificationTimestamp. The frequency of the verifications is such that this optimization is just not going to be particularly useful.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

nvb committed:
storage: Only scan over replicated data during GC queue processing

This reduces the number of keys that need to be processed in the
`gcQueue`, but means that we can no longer set the last verification
timestamp at the end of the GC queue processing.

nvanbenschoten commented:
Review status: 0 of 13 files reviewed at latest revision, 5 unresolved discussions.

---

*[storage/gc_queue.go, line 153 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCISF9lv91JoOmmvNHD-r1-153:-KCSRm8BwckoU-jO2ycf:1621646280) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf51391dcda579fe83757e32ef9443ac0b32a176/storage/gc_queue.go#L153)):*
SGTM. Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

tamird commented:
nit: prefer `GC queue" to the mix of "GCQueue" and "GC Queue" currently in the commit message

LGTM

---

Reviewed 1 of 6 files at r1, 1 of 8 files at r2, 11 of 11 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/replica_data_iter_test.go, line 119 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSQy2N9gFN5pR0F5qI:-KCSQy2N9gFN5pR0F5qJ:1306027387) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fa9e9e0a09428641647fabcbcba1de9e1ccf6/storage/replica_data_iter_test.go#L119)):*
should there be a unit test in this file that exercises the `true` case?

---

*[storage/replica_raftstorage.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCISF9lv91JoOmmvNHG-r1-491:-KCSQh6sEW7KdAAt_ze_:-1373426569) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf51391dcda579fe83757e32ef9443ac0b32a176/storage/replica_raftstorage.go#L491)):*
Can you preserve some of ben's comment as a comment in the code? or at least whichever part of it's absence initially made you pass `true` here.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCSRPy_yV4Rn__XsH85:-1031323808) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
how come this changed? just because general sketchiness of this pattern?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-:-KCSQQ5pDcNJFucnwHbp:-698134827)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 11 of 13 files reviewed at latest revision, 7 unresolved discussions.

---

*[storage/replica_data_iter_test.go, line 119 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSQy2N9gFN5pR0F5qI:-KCSXPkUmkGH_4mRGmUi:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/290fa9e9e0a09428641647fabcbcba1de9e1ccf6/storage/replica_data_iter_test.go#L119)):*
Done.

---

*[storage/replica_raftstorage.go, line 491 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCISF9lv91JoOmmvNHG-r1-491:-KCST18rd5wiqj974CCI:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/bf51391dcda579fe83757e32ef9443ac0b32a176/storage/replica_raftstorage.go#L491)):*
Done.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCST6IHyVJ2ctJFvxZe:-597109471) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
I believe the old way would allocate when taking the address of the `ent` copy. Plus having everything in one line that wraps seems less clean than separating them out into a few smaller lines.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCSZLLsOnm7gSJ4pYRm:-2117598496) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
Yeah, it might allocate, good point. While you're here you can get rid of the `ent` local and inline the `if err :=`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: 12 of 13 files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCS_OFsLgXwDS5p-BLK:-2048242486) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
Inlined the `if err :=`, even though it doesnt look like we use that pattern very much here. Maybe @bdarnell doesnt like it.

And why remove the local. We use it twice.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCSa73vE7knL3OLdimP:1308799193) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
oh, didn't see the `ent.Index`. carry on

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCSa1AnTW4AtLFU1mp4:-459751168) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
I don't have any objection to `if err :=`; it's just that most of the raft code predates the point where we started caring about that.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[storage/replica_raftstorage.go, line 422 \[r3\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960#-KCSRPy_yV4Rn__XsH84:-KCSaSNd4E0LdaRXGX7S:-391320432) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f89921cccdedbd564148fe068ade4f991256826/storage/replica_raftstorage.go#L422)):*
The code also has a number of long function calls that get wrapped, which certainly don't look as good with the `if err :=` pattern. I'm fine with either.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/4960)*
<!-- Sent from Reviewable.io -->
